### PR TITLE
[transactions3] Reworking CheckAndSetCompatibilities

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- Have a multi-dimensional way of expressing whether a KVS is able to perform the check-and-set (CAS) operation, and how well it can or cannot do so.

**Implementation Description (bullets)**:
- Make CASC have three booleans (supported, detailOnFailure, consistentOnFailure)
- Provide suitable interfaces to ensure users only build sensible configurations

**Testing (What was existing testing like?  What have you done to improve it?)**:
- Refined existing tests for CASC, and patched ones where necessary.

**Concerns (what feedback would you like?)**:
- I decided to make `unsupported()` _throw_ if one attempts to query the more precise detailed methods. Is this overly aggressive?

**Where should we start reviewing?**: CheckAndSetCompatibility itself

**Priority (whenever / two weeks / yesterday)**: today or tomorrow would be ideal

### CODE REVIEW

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/CheckAndSetCompatibility.java [hunk 1]
```diff
@@ -16,64 +16,95 @@
...
+
+    static CheckAndSetCompatibility unsupported() {
+        return Unsupported.INSTANCE;
+    }
+
+    static ImmutableSupported.Builder supportedBuilder() {
+        return ImmutableSupported.builder();
+    }
+
+    enum Unsupported implements CheckAndSetCompatibility {
```
Comment (gmaretic) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/CheckAndSetCompatibility.java:82]: nice!

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/KeyValueService.java [hunk 2]
```diff
@@ -258,14 +258,13 @@ void multiPut(Map<TableReference, ? extends Map<Cell, byte[]>> valuesByTable, lo
...
-        return getCheckAndSetCompatibility() != CheckAndSetCompatibility.NO_DETAIL_CONSISTENT_ON_FAILURE;
+        return getCheckAndSetCompatibility().supportsCheckAndSetOperations();
     }
 
     /**
      * Get the {@link CheckAndSetCompatibility} that this {@link KeyValueService} exhibits.
      *
-     * This method should be consistent with {@link KeyValueService#supportsCheckAndSet()} - this method should
-     * return {@link CheckAndSetCompatibility#NO_DETAIL_CONSISTENT_ON_FAILURE} if and only if that method returns false.
+     * This method should be consistent with {@link KeyValueService#supportsCheckAndSet()}.
```
Comment (gmaretic) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/KeyValueService.java:267]:
```suggestion
     * This method must be consistent with {@link KeyValueService#supportsCheckAndSet()}.
```

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/CheckAndSetCompatibility.java [hunk 3]
```diff
@@ -16,64 +16,95 @@
...
+        INSTANCE;
+
+        @Override
+        public boolean supportsCheckAndSetOperations() {
+            return false;
+        }
+
+        @Override
+        public boolean supportsDetailOnFailure() {
+            throw new SafeIllegalStateException(
```
Comment (gmaretic) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/CheckAndSetCompatibility.java:92]: Is this illegal state or illegal argument? Maybe even UnsupportedOperation tbh?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java [hunk 4]
```diff
@@ -431,7 +431,10 @@ private void putInternal(
 
     @Override
     public CheckAndSetCompatibility getCheckAndSetCompatibility() {
-        return CheckAndSetCompatibility.SUPPORTS_DETAIL_NOT_CONSISTENT_ON_FAILURE;
+        return CheckAndSetCompatibility.supportedBuilder()
+                .supportsDetailOnFailure(true)
+                .consistentOnFailure(true)
```
Comment (gmaretic) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java:436]: false to match Cassandra for tests?

**Review by gmaretic — COMMENTED** [review]
I like it overall, maybe would have even gone with Derive4J but this is nice enough

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/CheckAndSetCompatibility.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/CheckAndSetCompatibility.java:92]: Let's use UnsupportedOperation, yeah.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java (hunk 4 above)
Comment (jeremyk-91) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java:436]: Sure. It's kind of horrible and could be argued that the better way to do that is to have a `TestingInMemoryKeyValueService`, but that's probably too much work.

**Pull request merged**